src/main/java/com/ethanyidong/bunny/cmd/ExecutableCommand.java

```diff
@@ -0,0 +1,60 @@
+package com.ethanyidong.bunny.cmd;
+
+import com.ethanyidong.bunny.BunnySession;
+import com.ethanyidong.bunny.arg.CommandValidator;
+import com.ethanyidong.bunny.arg.InvalidArgumentException;
+import com.ethanyidong.bunny.arg.InvalidCommandException;
+
+public abstract class ExecutableCommand {
+    private static ExecutableCommand fromCommandString(String commandString) {
+        switch (commandString) {
+        case "bye":
+            return new ByeCommand();
+        case "list":
+            return new ListCommand();
+        case "mark":
+            return new MarkCommand(true);
+        case "unmark":
+            return new MarkCommand(false);
+        case "todo":
+            return new AddTodoCommand();
+        case "deadline":
+            return new AddDeadlineCommand();
+        case "event":
+            return new AddEventCommand();
+        case "delete":
+            return new DeleteTaskCommand();
+        default:
+            return null;
+        }
+    }
+
+    public static ExecutableCommand validateAndParse(BunnySession bunny, TokenizedCommand command)
+            throws InvalidCommandException {
+        ExecutableCommand newExecutableCommand = ExecutableCommand.fromCommandString(command.getCommand());
+
+        if (newExecutableCommand == null) {
+            throw new InvalidCommandException("command", new InvalidArgumentException("is unrecognized"));
+        }
+        newExecutableCommand.validateCommand(bunny, command);
+        newExecutableCommand.parseArguments(bunny, command);
+        return newExecutableCommand;
+    }
+
+    protected CommandValidator[] validators() {
```

Comment:
The name of a function should describe what it does. As such, `validators` does not give the reader a clear picture of what the function is supposed to do/return. Perhaps consider renaming it to `getValidators` or something along those lines. [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-name-to-explain)